### PR TITLE
8323672: Suppress unwanted autoconf added flags in CC and CXX

### PR DESCRIPTION
**Pull request by TheShermanTanker**
[JDK-8323008](https://bugs.openjdk.org/browse/JDK-8323008) reports unwanted autoconf flags being added to the command line, and solves the issue by filtering out the added flags by force. This is not optimal however, as doing so leaves the autoconf message that the flags were added still displaying during the configure process, which is confusing. Instead, setting a couple of fields to disable the autoconf internals responsible for the unwanted flag is a cleaner solution

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323672](https://bugs.openjdk.org/browse/JDK-8323672): Suppress unwanted autoconf added flags in CC and CXX (**Bug** - P4)

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17401/head:pull/17401` \
`$ git checkout pull/17401`

Update a local copy of the PR: \
`$ git checkout pull/17401` \
`$ git pull https://git.openjdk.org/jdk.git pull/17401/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17401`

View PR using the GUI difftool: \
`$ git pr show -t 17401`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17401.diff">https://git.openjdk.org/jdk/pull/17401.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17401#issuecomment-1889429071)

### CODE REVIEW

File: make/autoconf/toolchain.m4 [hunk 1]
```diff
@@ -352,6 +352,14 @@ AC_DEFUN_ONCE([TOOLCHAIN_PRE_DETECTION],
   # autoconf magic only relies on PATH, so update it if tools dir is specified
   OLD_PATH="$PATH"
 
+  # autoconf can ass unwanted flags to CC and CXX based on what it deems to be
```
Comment (magicus) [inline make/autoconf/toolchain.m4:355]:
```suggestion
  # autoconf can add unwanted flags to CC and CXX based on what it deems to be
```

**Comment by magicus** [comment]
I'm kind of on the fence for this one. On one hand filtering out stuff is really hacky, but so is setting internal undocumented variables. We have done some meddling with internal variable in a few places before, but I'd really like to keep that to a minumum. I'm not sure this is an improvement. I'd like to get input from other build folks on this.

**Comment by TheShermanTanker** [comment]
Hmm, wonder what it'd be like if we rolled our own AC_PROG_CC?

**Comment by TheShermanTanker** [comment]
I was thinking earlier this morning if I could roll a custom AC\_PROG\_CC for the JDK, since it could solve the issues that the current AC\_PROG\_CC and AC\_PROG\_CXX (including the addition of flags that we don't want). This way we could avoid hacks and messing with autoconf internals related to the problems that AC\_PROG\_CC presents (such as the one present here). Any thoughts on this?

**Comment by magicus** [comment]
The current use of AC_PROG_CC does not fit very well with out needs. It does a lot of testing that is not needed for us (and at least on Windows, this considerably slows down configure). So it would make sense to replace it with something that only does what we need. I have thought of this before, even without the context of this bug.

However, I have not pushed in this direction yet, partly because of lack of time, but also partly since I was afraid (well-grounded or not) that this was doing a lot of "autoconf black magic" that are needed elsewhere. 

But if you can take the time to analyze and understand what the vanilla AC_PROG_CC macro actually does, and extract just the parts that are relevant for us, and argue convincingly that the parts we're leaving out is not going to affect us, then I think it would be good. 

The more I've worked with autoconf, the less happy I've become with all its magic stuff, unchangable premises and bad fit for our use case.

**Comment by TheShermanTanker** [comment]
What fun, I just found out that omitting the call to AC\_PROG\_CC and AC\_PROG\_CXX has no effect, and they're _still_ called magically from somewhere by autoconf even if you don't call them. I'm genuinely at a loss for words :)

**Comment by magicus** [comment]
Then it sounds like this is not a viable approach. 

Also, check your inbox for a personal invitation to The "wtf-autoconf-rly?" Club. ;-)

**Comment by TheShermanTanker** [comment]
> Then it sounds like this is not a viable approach.

There's definitely gotta be a way to make autoconf do what we want that's out there, I'll work on it and post any findings back here. I've just found out that the implicit calls to AC\_PROG\_CC are due to AC\_PROG\_CPP calling AC\_REQUIRE on AC\_PROG\_CC for instance



> Also, check your inbox for a personal invitation to The "wtf-autoconf-rly?" Club. ;-)

I think I'm already a long time member of that club by this point...
(Sometimes I wonder whether Java could've gone the way V8 went and used Python or Java itself as the build system instead, oh well)

**Comment by TheShermanTanker** [comment]
> > (Sometimes I wonder whether Java could've gone the way V8 went and used Python or Java itself as the build system instead, oh well)
> 
> While I think we're stuck with make for the long run, a long-term goal for me has been (and still is) to replace autoconf with a system written in Java. Fortunately, we have a well defined interface: The user runs `configure` with a set of well-know arguments, we check for a bunch of requirements and the validity of the requested build configuration, and generates a well-defined spec.gmk file as a result. This tight encapsulation means it is possible to make a drop-in replacement in Java. The challenge is "just" to make sure we don't loose any functionality, nor skip over any of the odd fixes and adaptation to real-world build systems that the current configure script has collected over the years.

That sounds like a huge potential improvement, at least to me. I've also just come across JEP\-138 (Fascinating cache of JDK history within that JEP! Never knew the JDK used to only work with make), so it's interesting to see if autoconf can be phased out after its introduction 9 years ago. If I understand you correctly, this new hypothetical system would replace autoconf as such?

configure -> autoconf -> compiled configure -> make
to
configure -> compile Java -> Java configure program -> make

**Comment by erikj79** [comment]
> What fun, I just found out that omitting the call to AC_PROG_CC and AC_PROG_CXX has no effect, and they're _still_ called magically from somewhere by autoconf even if you don't call them. I'm genuinely at a loss for words :)

If I'm to guess, I think this is caused by AC_PROG_* being defined as AC_DEFUN_ONCE and every other autoconf macro that uses compilers for checks will implicitly call AC_PROG_* first if it hasn't been called yet. My understanding of these implicit calls is to make a more declarative programming style possible.

**Comment by TheShermanTanker** [comment]
> > What fun, I just found out that omitting the call to AC_PROG_CC and AC_PROG_CXX has no effect, and they're _still_ called magically from somewhere by autoconf even if you don't call them. I'm genuinely at a loss for words :)
> 
> If I'm to guess, I think this is caused by AC_PROG_* being defined as AC_DEFUN_ONCE and every other autoconf macro that uses compilers for checks will implicitly call AC_PROG_* first if it hasn't been called yet. My understanding of these implicit calls is to make a more declarative programming style possible.

It's not so much the AC\_DEFUN\_ONCE, but actually due to other autoconf macros calling AC\_REQUIRE([AC\_PROG\_CC]) and AC\_REQUIRE([AC\_PROG\_CXX]), for instance our calls to AC\_PROG\_CPP and AC\_PROG\_CXXCPP will cause AC\_PROG\_CC and AC\_PROG\_CXX respectively to be implicitly called

**Comment by TheShermanTanker** [comment]
I found out that the issue of having AC\_PROG\_CC and AC\_PROG\_CXX being called by AC\_REQUIRE can be solved by directly overwriting them in util.m4, rather than providing our own UTIL\_PROG\_CC and so on. Unsure how satisfactory this is, but I have here a stripped down version of both that omits:

For AC\_PROG\_CC
- The default check from a list of compilers if the compilers to search for list is empty, this now means AC\_PROG\_CC has to always be called with the argument at all times. We always do this, so shouldn't be a problem
- The check whether the compiler supports GNU C. This shouldn't be relevant
- The ill fated check that adds -std=gnu11. This was the one that the other issue wanted to get rid of

For AC\_PROG\_CXX
- Similarly, the default check from a list of compilers if the compilers to search for list is empty, and the check for CCC if it is set (We don't use CCC as fair as I can tell), this now means AC\_PROG\_CXX has to always be called with the argument at all times. We again always do this, so shouldn't be a problem
- The check whether the compiler supports GNU C++. This shouldn't be relevant
- The ill fated check that adds -std=gnu++11. This is also the one that the other issue wanted to get rid of

Notably, I left the check that checks if the compiler accepts -g in there, as I was unsure if we need it or not. I also tried removing the check for the object file extension, but apparently autoconf needs that to function properly. The exe file extension check is misleadingly named, because in actuality all of AC\_PROG\_CC and AC\_PROG\_CXX's functionality is in it,and it's also likely required for autoconf to work properly

**Comment by erikj79** [comment]
> I found out that the issue of having AC_PROG_CC and AC_PROG_CXX being called by AC_REQUIRE can be solved by directly overwriting them in util.m4, rather than providing our own UTIL_PROG_CC and so on. Unsure how satisfactory this is, but I have here a stripped down version of both that omits:
> 
> For AC_PROG_CC
> 
> * The default check from a list of compilers if the compilers to search for list is empty, this now means AC_PROG_CC has to always be called with the argument at all times. We always do this, so shouldn't be a problem
> * The check whether the compiler supports GNU C. This shouldn't be relevant
> * The ill fated check that adds -std=gnu11. This was the one that the other issue wanted to get rid of
> 
> For AC_PROG_CXX
> 
> * Similarly, the default check from a list of compilers if the compilers to search for list is empty, and the check for CCC if it is set (We don't use CCC as fair as I can tell), this now means AC_PROG_CXX has to always be called with the argument at all times. We again always do this, so shouldn't be a problem
> * The check whether the compiler supports GNU C++. This shouldn't be relevant
> * The ill fated check that adds -std=gnu++11. This is also the one that the other issue wanted to get rid of
> 
> Notably, I left the check that checks if the compiler accepts -g in there, as I was unsure if we need it or not. I also tried removing the check for the object file extension, but apparently autoconf needs that to function properly. The exe file extension check is misleadingly named, because in actuality all of AC_PROG_CC and AC_PROG_CXX's functionality is in it,and it's also likely required for autoconf to work properly

This does look like an interesting solution and far smaller than I had anticipated. The risk is that we start depending on autoconf internals that may change in future versions. We would also need to validate this solution against all currently supported versions. At least I'm assuming that macros starting with `_` aren't part of the public API.

**Comment by magicus** [comment]
I agree with Erik, this looks interesting. I also like the benefit of transparency that it brings, breaking down the black box of autoconf a bit more. 

In contrast with Erik I'm not so worried about future breakage. Autoconf has basically stalled in development since decades. If someone were to pick up development again (like what happened with GNU Make) we will surely see signs of activity long before this breaks. And if that happens, we can just file a bug on autoconf instead of trying to work around the broken behaviour, right? 🤷 

I would like to take this PR for a spin on a bunch of machines/environments, though, to see that it does not fail obviously on any of them.

**Comment by TheShermanTanker** [comment]
Haha, thanks both. Though one thing has been annoying me ever since I wrote this, it's the fact that no "AC\_UNDEF" exists to erase macros defined by AC\_DEFUN. Overwriting macros like this, especially macros that are part of the "autoconf standard library", is a little bit painful to me. On the other hand, not like there is another way either...
(I guess m4\_undefine exists, but that seems to be more for macros defined with m4\_define, though I have tried using it on macros defined by AC\_DEFUN and it _did_ work)

**Comment by erikj79** [comment]
> In contrast with Erik I'm not so worried about future breakage. Autoconf has basically stalled in development since decades. If someone were to pick up development again (like what happened with GNU Make) we will surely see signs of activity long before this breaks. And if that happens, we can just file a bug on autoconf instead of trying to work around the broken behaviour, right? 🤷

If you are ok with this, then I am too, just wanted to be thorough with possible concerns. I would just like to point out that it was a change in autoconf behavior that triggered this whole ordeal in the first place, so saying nothing is happening doesn't exactly seem true. The way I see it, if they change one of the sub/internal macros in an incompatible way, then we will just have to keep inlining the old versions of such macros in our configure script to work around it.

I still think it would be prudent to verify this patch with all the currently accepted versions of autoconf (2.69, 2.70, 2.71, 2.72).

**Comment by erikj79** [comment]
> Haha, thanks both. Though one thing has been annoying me ever since I wrote this, it's the fact that no "AC_UNDEF" exists to erase macros defined by AC_DEFUN. Overwriting macros like this, especially macros that are part of the "autoconf standard library", is a little bit painful to me. On the other hand, not like there is another way either... (I guess m4_undefine exists, but that seems to be more for macros defined with m4_define, though I have tried using it on macros defined by AC_DEFUN and it _did_ work)

I don't think it's strange that they aren't duplicating existing m4 functionality. If you want to undefine macros you are already moving outside of the intended autoconf APIs so having to reach for direct m4 functionality would be expected.

**Comment by magicus** [comment]
>  I would just like to point out that it was a change in autoconf behavior that triggered this whole ordeal in the first place

I'm not entirely sure that is the case. It might as well have been a latent bug that was triggered by changes in the environment, or the compiler, or something like that. I don't think we ever got to the bottom of what caused it.

**Comment by magicus** [comment]
Correction: I checked the history, and you are correct. It was a change in autoconf version from 2.71 to 2.72 that caused this.

**Comment by magicus** [comment]
I have verified that this works fine in the Oracle internal CI.

Erik's point still stands:
> I still think it would be prudent to verify this patch with all the currently accepted versions of autoconf (2.69, 2.70, 2.71, 2.72).

I think the easiest way to achieve this is to checkout the autoconf at these versions, and build it yourself. Iirc it was quite easy to build autoconf (anything else would be a shame and very bad PR for the project! :-o).

I'm hoping you are willing to do this, since I believe this is a superior solution and I'd like to see it in mainline. (Otherwise, let me know and I'll try to squeeze in doing it.)

**Comment by magicus** [comment]
> That said, should I test autoconf on all platforms too?

No, that seems over the top. For this purpose, we must assume that all autoconf on all platforms behave the same.

**Comment by magicus** [comment]
@TheShermanTanker I think this is essentially done. All that needs are the testing. Can you do it or do you want help with it?

**Comment by TheShermanTanker** [comment]
> @TheShermanTanker I think this is essentially done. All that needs are the testing. Can you do it or do you want help with it?

I'm able to do so myself, I've just been busy with university so far. I've only tested 2.71 and 2.72 though, 2.69 and 2.70 might still need testing. I'm still unhappy that I can't properly undefine the macros, as a side tangent

**Comment by magicus** [comment]
Keep it for a while more, bot. I'll try to some testing of this on 2.69 and 2.70 if Julian is unable to.

**Comment by magicus** [comment]
Ok. I have now tested this patch with autoconf 2.69 and 2.70, and afaict it works just fine. So this is finally ready to integrate. @TheShermanTanker Will you have the honors?

**Comment by TheShermanTanker** [comment]
Sorry for essentially having you do the testing on my behalf. I'm still very unhappy that I have to overwrite an AC\_DEFUN function like this, but there doesn't seem to be any other way. I'll fix the conflicts first though

**Review by magicus — APPROVED** [review]

**Pull request closed without merging**